File: Dynamic_Programming/7_Oct_2019_nth_Catalan_number/7_Oct_2019_nth_Catalan_number.cpp
//naive recursive catalan numbers

#include<bits/stdc++.h>
using namespace std;

int catalan(int n)
{
    int i;
    //base case 
    if(n<=1){
        return 1;
    }

    //catalan(n) is sum of catalan(i)*catalan(n-i-1) 
    int res = 0 ;
    for(i = 0; i < n; i++){
        res  += catalan(i)*catalan(n-i-i);
    }
    return res;
}

//driver program
int main(){
    for (int i = 0; i< 10; i++){
        cout << catalan(i) << " ";
    }
    return 0;
}

File: Dynamic_Programming/7_Oct_2019_nth_Catalan_number/7_Oct_2019_nth_Catalan_number.py
#naive recursive catalan

def catalan(n):
    #base case
    if(n<=0):
        return 1
    
    # Catalan(n) is the sum of catalan(i)*catalan(n-i-1)
    res = 0
    for i in range(n):
        res += catalan(i) * catalan(n-i-1)
    
    return res

#driver 
for i in range(11):
    print(" {}".format(catalan(i)))